### DISCUSSION TITLE
Error when removing and adding trainable pipes

### DISCUSSION THREAD

**Discussion by jonabaa** [explosion/spaCy#9047] [Help: Coding & Implementations] [answered]
I want to load a pre-trained pipeline (en_core_web_lg) with all pipes excluded and then add the pipes I need. I use spacy 3.0.6. As a test I run the following code:

```
import spacy
from spacy.pipeline.tok2vec import DEFAULT_TOK2VEC_MODEL

nlp = spacy.load("en_core_web_lg", exclude=["tok2vec", "tagger", "parser", "ner", "attribute_ruler", "lemmatizer"])

config = {"model": DEFAULT_TOK2VEC_MODEL}
nlp.add_pipe("tok2vec", config=config)

text = "This is a sample text."
doc = nlp(text)
```

However, I get the following error:

```
[...]
  File "/home/jab/miniconda3/envs/xait3.8_spacy3.0/lib/python3.8/site-packages/thinc/model.py", line 212, in get_param
    raise KeyError(
KeyError: "Parameter 'E' for model 'hashembed' has not been allocated yet."

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "discussion_spacy.py", line 10, in <module>
    doc = nlp(text)
  File "/home/jab/miniconda3/envs/xait3.8_spacy3.0/lib/python3.8/site-packages/spacy/language.py", line 998, in __call__
    raise ValueError(Errors.E109.format(name=name)) from e
ValueError: [E109] Component 'tok2vec' could not be run. Did you forget to call `initialize()`?

```
Hence it looks like I should run the initialize method on the tok2vec-object. But I just want to use the already trained model, not to initialize the component for training. I think there is some detail I am missing out on here.

**Comment by polm** [comment, marked as answer]
When you call `nlp.add_pipe("something")`, a fresh `something` is created from scratch. If you look at the [`add_pipe` docs](https://spacy.io/api/language#add_pipe) you can see that you can use `source` to specify the name of a pipeline to take the component from. (Note that when listeners are involved, as with the default pipelines, this can be more complicated.)

> I want to load a pre-trained pipeline (en_core_web_lg) with all pipes excluded and then add the pipes I need. 

Why do you want to do this? It would likely be easier and faster to load the pipeline with the components you will use simply disabled, and then enable them as needed. Can you give me some more perspective on your use case?

**Reply by jonabaa** [reply]
Thanks for your answer polm!

First some context: My pipeline contains quite a lot (a growing number) of pipes and I wish to group them together into "super-pipes". That is, I have created a custom (abstract) pipe called _SuperPipe_. Its purpose is to group together a selection of pipes into its own pipeline, run this when called and filter the results. An example: I wish to use three different pipes to set entity labels (including EntityRecognizer), but still want this to work like one pipe in the pipeline of the Language object.

With this context in mind, I wish to start out with an empty pipeline and then add my custom super-pipes containing their own set of pipes.

Regarding my previous post: I thought the default config for Language.add_pipe(<factory_name>) would set up a ready to use pipe with an already trained model. I understand now that it rather gives you a new pipe with a blank model ready for initialization and training?

To summarize: My goal is to set up the pipes from the default pipeline ("tagger", "parser", "ner" and so on) using Language.create_pipe and add them to my custom pipelines (owned by the SuperPipe instances). To me it looks like I have to pass a custom config with create_pipe to achieve this? It also seems like I have to deserialize the model myself as this is not done by create_pipe? Finally I have to take care of the listeners connecting them to tok2vec/transformer but this seems to also be done via the config. Anything else that needs to be set up? If you have any suggestions for reading on this topic, or parts of the spacy code I should look into, please let me know.

**Reply by polm** [reply]
Thanks for providing more perspective on your use case. Can you clarify what you mean by "work like one pipe" though? Are you controlling how the components overwrite each other or...? I haven't heard of anyone doing that before, and if you want to expand on why it's helpful that'd be good to know, in case it might be a useful general feature.

> Regarding my previous post: I thought the default config for Language.add_pipe(<factory_name>) would set up a ready to use pipe with an already trained model. I understand now that it rather gives you a new pipe with a blank model ready for initialization and training?

Yes, the components are blank by default. If you want to load from an existing pipeline you have to specify that. 

> To me it looks like I have to pass a custom config with create_pipe to achieve this? 

Not necessarily. If you look at [the docs on sourcing components][sourcing], it can be pretty simple, but it depends on if you have the same tok2vec and word vectors around. If you have your own EntityRecognizer the tok2vec has probably changed, so you'll need to look at details on [replacing listeners](https://spacy.io/usage/training#:~:text=shared%20tok2vec%20listener%20layer). If you have changed vectors... I'd need to look up how that works.

[sourcing]: https://spacy.io/usage/processing-pipelines#sourced-components

> It also seems like I have to deserialize the model myself as this is not done by create_pipe? 

See the example in the sourcing docs, that does deserialize the source model first.

> With this context in mind, I wish to start out with an empty pipeline and then add my custom super-pipes containing their own set of pipes.

Assuming they are specified like normal components (which I don't think should be a problem?), you can specify your whole architecture, including source pipes and superpipes and order and options, in the config. If you train the components separately and only put them together later you can use [`assemble`][assemble] to load from a config.

[assemble]: https://spacy.io/api/cli#assemble
